pkg/controller/masteruserrecord/masteruserrecord_controller.go

```diff
@@ -303,18 +303,31 @@ func (r *ReconcileMasterUserRecord) deleteUserAccount(logger logr.Logger, target
 	// Get the User associated with the UserAccount
 	userAcc := &toolchainv1alpha1.UserAccount{}
 	namespacedName := types.NamespacedName{Namespace: memberCluster.OperatorNamespace, Name: name}
-	err = memberCluster.Client.Get(context.TODO(), namespacedName, userAcc)
-	if err != nil {
+	if err = memberCluster.Client.Get(context.TODO(), namespacedName, userAcc); err != nil {
 		if errors.IsNotFound(err) {
 			return nil
 		}
 		return err
 	}
+
+	if coputil.IsBeingDeleted(userAcc) {
+		return fmt.Errorf("UserAccount deletion in progress")
+	}
+
 	if err := memberCluster.Client.Delete(context.TODO(), userAcc); err != nil {
 		return err
 	}
 	counter.DecrementUserAccountCount(logger, targetCluster)
-	return nil
+
+	// Verify the UserAccount is deleted
+	userAcc = &toolchainv1alpha1.UserAccount{}
+	if err = memberCluster.Client.Get(context.TODO(), namespacedName, userAcc); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	return fmt.Errorf("waiting for UserAccount deletion") // return error to reconcile again and wait for deletion
```

Comment:
I don't think that this is really useful - the deletion won't be done for sure at this point in the time. The reason is that the member controllers have to delete many other resources which takes some time.
I would just return with postponed reconciliation for example by 5-10 seconds.

TBH, there will be some additional reconciliations triggered by `UserAccount_Status_Controller` based on the changed status in UserAccount - the longer delay for the next reconciliation is just to be sure that we don't miss the final deletion.

Reply:
Yeah, I was thinking about this too but wasn't sure whether it could be deleted quickly or not. If it won't in most cases then I agree, we can just remove it.

Reply:
Done in https://github.com/codeready-toolchain/host-operator/pull/352/commits/e9a3765ce783f8eb8561e95d79636723ef9fc666